src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs

```diff
@@ -1286,6 +1286,52 @@ public static class PipelineStageDefinitionBuilder
             return SortByCount(new ExpressionAggregateExpressionDefinition<TInput, TValue>(value, translationOptions));
         }
 
+        /// <summary>
+        /// Creates a $unionWith stage.
+        /// </summary>
+        /// <typeparam name="TInput">The type of the input documents.</typeparam>
+        /// <typeparam name="TWith">The type of the with collection documents.</typeparam>
+        /// <param name="withCollection">The with collection.</param>
+        /// <param name="withPipeline">The with pipeline.</param>
+        /// <returns>The stage.</returns>
+        public static PipelineStageDefinition<TInput, TInput> UnionWith<TInput, TWith>(
+            IMongoCollection<TWith> withCollection,
+            PipelineDefinition<TWith, TInput> withPipeline = null)
+        {
+            Ensure.IsNotNull(withCollection, nameof(withCollection));
```

Comment:
Let's move the check for a missing withPipeline here. We don't want to wait until Render is called to get the exception.

    if (withPipeline == null && typeof(TWith) != typeof(TInput))
    {
        throw new ArgumentException("The withPipeline cannot be null when TWith != TInput. A pipeline is required to transform the TWith documents to TInput documents.", nameof(withPipeline));
    }